lib/std/defers.nim

```diff
@@ -0,0 +1,12 @@
+template deferScoped*(cleanup, body) =
+  ## Similar to builtin `defer` statement, but operates on a scope `body`,
+  ## whereas `defer` operates until end of block it's defined in.
+  ## Implemented as: `try: body ... finally: cleanup`
+  runnableExamples:
+    from std/strutils import contains
+    let a = open(currentSourcePath)
+    deferScoped: close(a) # immediately follows code that needs cleanup
+    do:
```

Comment:
Oh please. This is even worse than ``defer``. There is no visual indication that the ``do:`` block belongs to the ``deferScoped``.

Reply:
@Araq 
> There is no visual indication that the do: block belongs to the deferScoped.

you don't have to use `do`, eg this works, and shows that visual indication:
```nim
echo "before"
deferScoped(echo "cleanup"):
  echo "inner"

echo "before"
deferScoped(echo "cleanup"):
  echo "inner"
```

but `do` is part of the language and users should learn how it works (and you could use same argument for other API's using do)


**The whole point is it doesn't escape the template/macro scope it's called from, unlike `defer`**

```nim
template fn():
  prelude()
  deferSCoped: cleanup()
  do:
    inner()

fn()
# `cleanup` called before the next statement
fn()
# `cleanup` called before the next statement
```